File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
class MedianFinder {
public:
    priority_queue<int> max_heap;
    priority_queue<int, vector<int>, greater<int>> min_heap;

    MedianFinder() {}

    void addNum(int num) {
        // elements ko both heap me daal diyea hain
        if (max_heap.empty() || num < max_heap.top()) {
            max_heap.push(num);
        } else {
            min_heap.push(num);
        }
        // always maintain max_heap size > than min heap sieze ya fir equal ...

        if (abs((int)max_heap.size() - (int)min_heap.size()) > 1) {
            min_heap.push(max_heap.top());
            max_heap.pop();
        } else if (max_heap.size() < min_heap.size()) {
            max_heap.push(min_heap.top());
            min_heap.pop();
        }
    }

    double findMedian() {
        if (max_heap.size() == min_heap.size()) {
            return (double) (max_heap.top() + min_heap.top())/2;
        } else {
            return max_heap.top();
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.max_heap = [] #left part of num will store
        self.min_heap = [] #right part of num will store

    def addNum(self, num: int) -> None:
        heapq.heappush(self.max_heap , -num) # put num in max heap 
        largest_small = -heapq.heappop(self.max_heap)
        heapq.heappush(self.min_heap , largest_small)
        if(len(self.min_heap) > len(self.max_heap)):
            smallest_large = heapq.heappop(self.min_heap)
            heapq.heappush(self.max_heap , -smallest_large)

    def findMedian(self) -> float:
        if(len(self.max_heap) > len(self.min_heap)):
            return  -self.max_heap[0]
        else:
            small = -self.max_heap[0]
            large = self.min_heap[0]
        return (small + large) / 2
        
# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()